File: stgit/commands/sink.py
from stgit.argparse import keep_option, opt, patch_range
from stgit.commands.common import CmdException, DirectoryHasRepository, parse_patches
from stgit.lib import transaction

__copyright__ = """
Copyright (C) 2007, Yann Dirson <ydirson@altern.org>

This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License version 2 as
published by the Free Software Foundation.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program; if not, see http://www.gnu.org/licenses/.
"""

help = 'Send patches deeper down the stack'
kind = 'stack'
usage = ['[-t <target patch>] [-n] [--] [<patches>]']
description = """
This is the opposite operation of linkstg:float[]: move the specified
patches down the stack.  It is for example useful to group stable
patches near the bottom of the stack, where they are less likely to be
impacted by the push of another patch, and from where they can be more
easily committed or pushed.

If no patch is specified on command-line, the current patch gets sunk.
By default patches are sunk to the bottom of the stack, but the '--to'
option allows one to place them under any applied patch.

Sinking internally involves popping all patches (or all patches
including <target patch>), then pushing the patches to sink, and then
(unless '--nopush' is also given) pushing back into place the
formerly-applied patches."""

args = [patch_range('applied_patches', 'unapplied_patches')]
options = (
    [
        opt(
            '-n',
            '--nopush',
            action='store_true',
            short='Do not push the patches back after sinking',
            long="""
        Do not push back on the stack the formerly-applied patches.
        Only the patches to sink are pushed.""",
        ),
        opt(
            '-t',
            '--to',
            metavar='TARGET',
            args=['applied_patches'],
            short='Sink patches below the TARGET patch',
            long="""
        Specify a target patch to place the patches below, instead of
        sinking them to the bottom of the stack.""",
        ),
    ]
    + keep_option()
)

directory = DirectoryHasRepository()


def func(parser, options, args):
    """Sink patches down the stack.
    """
    stack = directory.repository.current_stack

    if options.to and options.to not in stack.patchorder.applied:
        raise CmdException('Cannot sink below %s since it is not applied' % options.to)

    if len(args) > 0:
        patches = parse_patches(args, stack.patchorder.all)
    else:
        # current patch
        patches = list(stack.patchorder.applied[-1:])

    if not patches:
        raise CmdException('No patches to sink')
    if options.to and options.to in patches:
        raise CmdException('Cannot have a sinked patch as target')

    applied = [p for p in stack.patchorder.applied if p not in patches]
    if options.to:
        insert_idx = applied.index(options.to)
    else:
        insert_idx = 0
    applied = applied[:insert_idx] + patches + applied[insert_idx:]
    unapplied = [p for p in stack.patchorder.unapplied if p not in patches]

    iw = stack.repository.default_iw
    clean_iw = (not options.keep and iw) or None
    trans = transaction.StackTransaction(stack, 'sink', check_clean_iw=clean_iw)

    try:
        trans.reorder_patches(applied, unapplied, iw=iw, allow_interactive=True)
    except transaction.TransactionHalted:
        pass
    return trans.run(iw)

File: src/cmd/sink.rs
// SPDX-License-Identifier: GPL-2.0-only

//! `stg sink` implementation.

use anyhow::{anyhow, Result};
use clap::{Arg, ArgMatches};

use crate::{
    argset,
    color::get_color_stdout,
    ext::RepositoryExtended,
    patch::{patchrange, LocationConstraint, PatchLocator, PatchName, PatchRange, RangeConstraint},
    stack::{InitializationPolicy, Stack, StackStateAccess},
    stupid::Stupid,
};

pub(super) const STGIT_COMMAND: super::StGitCommand = super::StGitCommand {
    name: "sink",
    category: super::CommandCategory::StackManipulation,
    make,
    run,
};

fn make() -> clap::Command {
    clap::Command::new(STGIT_COMMAND.name)
        .about("Move patches deeper in the stack")
        .long_about(
            "Move the specified patches down the stack.\n\
             \n\
             If no patch is specified on the command line, the current (topmost) patch \
             is sunk. By default, patches are sunk to the bottom of the stack, but the \
             '--to' option may be used to place them under any applied patch.\n\
             \n\
             Internally, sinking involves popping all patches to the bottom (or to the \
             target patch if '--to' is used), then pushing the patches to sink, and \
             then, unless '--nopush' is specified, pushing back any other formerly \
             applied patches.\n\
             \n\
             Sinking may be useful, for example, to group stable patches at the bottom \
             of the stack where they less likely to be impacted by the push of another \
             patch, and from where they can be more easily committed or pushed to \
             another repository.\n\
             ",
        )
        .arg(
            Arg::new("patchranges")
                .help("Patches to sink")
                .value_name("patch")
                .num_args(1..)
                .allow_hyphen_values(true)
                .value_parser(clap::value_parser!(PatchRange)),
        )
        .arg(
            Arg::new("nopush")
                .long("nopush")
                .short('n')
                .help("Do not push any patches back after sinking")
                .long_help(
                    "Do not push any formerly applied patches after sinking. \
                     Only the patches to sink are pushed.",
                )
                .action(clap::ArgAction::SetTrue),
        )
        .arg(
            Arg::new("target")
                .long("to")
                .short('t')
                .help("Sink patches below <target> patch")
                .long_help(
                    "Sink patches below <target> patch.\n\
                     \n\
                     Specified patches are placed below <target> instead of at the \
                     bottom of the stack.",
                )
                .value_name("target")
                .allow_hyphen_values(true)
                .value_parser(clap::value_parser!(PatchLocator)),
        )
        .arg(argset::keep_arg())
        .arg(argset::committer_date_is_author_date_arg())
}

fn run(matches: &ArgMatches) -> Result<()> {
    let repo = gix::Repository::open()?;
    let stack = Stack::from_branch(&repo, None, InitializationPolicy::AllowUninitialized)?;
    let stupid = repo.stupid();

    let nopush_flag = matches.get_flag("nopush");
    let keep_flag = matches.contains_id("keep");

    repo.check_repository_state()?;
    let statuses = stupid.statuses(None)?;
    statuses.check_conflicts()?;
    stack.check_head_top_mismatch()?;
    if !keep_flag {
        statuses.check_index_and_worktree_clean()?;
    }

    let opt_target: Option<PatchName> = matches
        .get_one::<PatchLocator>("target")
        .map(|loc| loc.resolve_name(&stack))
        .transpose()
        .map_err(|e| anyhow!("target: {e}"))?
        .map(|name| name.constrain(&stack, LocationConstraint::Applied))
        .transpose()
        .map_err(|e| match e {
            crate::patch::name::Error::PatchNotAllowed { patchname, .. } => {
                anyhow!("cannot sink below `{patchname}` since it is not applied")
            }
            _ => e.into(),
        })?;

    let patches: Vec<PatchName> =
        if let Some(range_specs) = matches.get_many::<PatchRange>("patchranges") {
            patchrange::resolve_names(&stack, range_specs, RangeConstraint::All)?
        } else if let Some(patchname) = stack.applied().last() {
            vec![patchname.clone()]
        } else {
            return Err(super::Error::NoAppliedPatches.into());
        };

    if let Some(target_patch) = &opt_target {
        if patches.contains(target_patch) {
            return Err(anyhow!(
                "target patch `{target_patch}` may not also be a patch to sink",
            ));
        }
    }

    let mut remaining_unapplied: Vec<PatchName> = stack
        .unapplied()
        .iter()
        .filter(|pn| !patches.contains(pn))
        .cloned()
        .collect();

    let mut remaining_applied: Vec<PatchName> = stack
        .applied()
        .iter()
        .filter(|pn| !patches.contains(pn))
        .cloned()
        .collect();

    let target_pos = if let Some(target_patch) = &opt_target {
        remaining_applied
            .iter()
            .position(|pn| pn == target_patch)
            .expect("already validated that target is applied")
    } else {
        0
    };

    let mut patches = patches;

    let (applied, unapplied) = if nopush_flag {
        let mut applied: Vec<PatchName> = Vec::with_capacity(target_pos + patches.len());
        applied.extend(remaining_applied.drain(0..target_pos));
        applied.append(&mut patches);
        let mut unapplied: Vec<PatchName> =
            Vec::with_capacity(remaining_applied.len() + remaining_unapplied.len());
        unapplied.append(&mut remaining_applied);
        unapplied.append(&mut remaining_unapplied);
        (applied, unapplied)
    } else {
        let mut applied: Vec<PatchName> =
            Vec::with_capacity(remaining_applied.len() + patches.len());
        applied.extend(remaining_applied.drain(0..target_pos));
        applied.append(&mut patches);
        applied.append(&mut remaining_applied);
        (applied, remaining_unapplied)
    };

    stack
        .setup_transaction()
        .use_index_and_worktree(true)
        .committer_date_is_author_date(matches.get_flag("committer-date-is-author-date"))
        .with_output_stream(get_color_stdout(matches))
        .transact(|trans| trans.reorder_patches(Some(&applied), Some(&unapplied), None))
        .execute("sink")?;

    Ok(())
}
